File: BOJ(Java 공부용)/2606.java
import java.io.*;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    static int[] gox = {0, 1, -1, 0};
    static int[] goy = {1, 0, 0, -1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());
        ArrayList<Integer>[] graph = new ArrayList[n + 1];
        boolean[] visited = new boolean[n + 1];
        for (int i = 0; i <= n; i++) graph[i] = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int here = Integer.parseInt(st.nextToken()), next = Integer.parseInt(st.nextToken());
            graph[here].add(next);
            graph[next].add(here);
        }
        dfs(1, graph, visited);
        int cnt = 0;
        for (int i = 1; i <= n; i++) {
            if (visited[i]) cnt++;
        }
        System.out.println(cnt - 1);
    }

    private static void dfs(int here, ArrayList<Integer>[] graph, boolean[] visited) {
        visited[here] = true;
        for (int i = 0; i < graph[here].size(); i++) {
            int next = graph[here].get(i);
            if (!visited[next]) dfs(next, graph, visited);
        }
    }
}

File: BOJ(Kotlin)/2606.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

val bw = BufferedWriter(OutputStreamWriter(System.out))
fun main(args : Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))

    val n = br.readLine().toInt()
    val m = br.readLine().toInt()
    val graph : Array<MutableList<Int>> = Array(n+1){ mutableListOf<Int>()}
    val visited = BooleanArray(n+1)
    for (i in 0 until m) {
        val st = StringTokenizer(br.readLine())
        val here = st.nextToken().toInt()
        val next = st.nextToken().toInt()
        graph[here].add(next)
        graph[next].add(here)
    }

    dfs(1, graph,visited)
    var cnt = 0
    for (i in 2..n) {
        if (visited[i]) cnt++
    }
    bw.write(cnt.toString())
    bw.flush()
    bw.close()
}

fun dfs(here : Int, graph : Array<MutableList<Int>>,visited : BooleanArray ) {
    visited[here] = true
    for(i in 0 until graph[here].size) {
        val next = graph[here].get(i)
        if (!visited[next]) {
            dfs(next,graph,visited)
        }
    }
}